From 89907f1f1a02e37dd7c093ed66b8e243c0643b8a Mon Sep 17 00:00:00 2001
From: Drew Davenport <ddavenport@chromium.org>
Date: Wed, 24 Apr 2019 17:13:19 -0600
Subject: [PATCH] st/mesa: Use GL_BGRA_EXT internal format for B8G8R8A8/B8G8R8X8

With a B8G8R8A8_UNORM / B8G8R8X8 image, when the app does the following:

glEGLImageTargetTexture2DOES(..)
glTexSubImage2D(..)

_mesa_gles_error_check_format_and_type(..) returns GL_INVALID_OPERATION (see
6ad9eb for details).  Let's specify the correct internal format up front to
correct this.

Change-Id: I93643d8ae9fa5178eb1468c1b1fedee70c2eb8f3
---
 src/mesa/state_tracker/st_cb_eglimage.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/mesa/state_tracker/st_cb_eglimage.c b/src/mesa/state_tracker/st_cb_eglimage.c
index f79df5a38ca5..3b88f7bad4a6 100644
--- a/src/mesa/state_tracker/st_cb_eglimage.c
+++ b/src/mesa/state_tracker/st_cb_eglimage.c
@@ -181,8 +181,11 @@ st_bind_egl_image(struct gl_context *ctx,
    mesa_format texFormat;
 
    /* map pipe format to base format */
-   if (util_format_get_component_bits(stimg->format,
-                                      UTIL_FORMAT_COLORSPACE_RGB, 3) > 0)
+   if (stimg->format == PIPE_FORMAT_B8G8R8A8_UNORM ||
+       stimg->format == PIPE_FORMAT_B8G8R8X8_UNORM)
+      internalFormat = GL_BGRA_EXT;
+   else if (util_format_get_component_bits(stimg->format,
+                                           UTIL_FORMAT_COLORSPACE_RGB, 3) > 0)
       internalFormat = GL_RGBA;
    else
       internalFormat = GL_RGB;
-- 
2.20.1

